src/Network/Transport/TCP.hs

```diff
@@ -830,6 +849,36 @@ handleConnectionRequest transport sock = handle handleException $ do
     rethrowIfAsync :: Maybe AsyncException -> IO ()
     rethrowIfAsync = mapM_ throwIO
 
+    probeIfValid :: LocalEndPoint -> RemoteEndPoint -> IO ()
+    probeIfValid ourEndPoint theirEndPoint = modifyMVar_ (remoteState theirEndPoint) $
+      \st -> case st of
+        RemoteEndPointValid
+          vst@(ValidRemoteEndPointState { remoteProbing = Nothing }) -> do
+            tid <- forkIO $ do
+              -- send probe
+              let params = transportParams transport
+              void $ tryIO $ System.Timeout.timeout
+                  (maybe (-1) id $ transportConnectTimeout params) $ do
+                sendMany sock [encodeInt32 ProbeSocket]
+                threadDelay maxBound
+              -- reset connection if this thread is not killed (i.e. the probe
+              -- ack does not arrive on time).
+              void $ forkIO $ do
+                tryCloseSocket (remoteSocket vst)
+                let closed = RemoteEndPointFailed $ userError "socket failed"
+                modifyMVar_ (remoteState theirEndPoint) $ \st' -> case st' of
+                  RemoteEndPointValid _ -> return closed
+                  RemoteEndPointClosing _ _ -> return closed
+                  _ -> return st'
+                removeRemoteEndPoint (ourEndPoint, theirEndPoint)
+                void $ tryIO $ createRemoteEndPoint params ourEndPoint
```

Comment:
Maybe reconnecting is not worth the complexity here. The other side will probably timeout before we try to reconnect.